### PR TITLE
Nested classes

### LINKED ISSUE #9: Introduce SDL classes
Instead of using pure C default SDL interface with functions like SDL_CreateWindow, SDL_DestroyWindow, etc, 
wrap them in C++ classes for better readability and memory management

### PR DESCRIPTION
**Pull request by Drone29**
added interfaces for window, renderer, surface, texture
closes #9

### CODE REVIEW

File: src/sdl_interface.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        virtual ~IRenderer() = default;
+        // Size
+        virtual std::tuple<size_t, size_t> get_output_size() const = 0;
+        // Modifiers
+        virtual IRenderer & clear() = 0;
+        virtual IRenderer & present() = 0;
+        virtual IRenderer & copy(std::unique_ptr<ITexture> & texture, const Rect *source, const Rect *dest) = 0;
+        // Nested objects
+        // Texture
+        virtual std::unique_ptr<ITexture> create_texture(ISurface &surface) = 0;
```
Comment (Drone29) [inline src/sdl_interface.h:71]: maybe it'd be better with std::unique_ptr<ISurface> &

File: src/sdl_interface.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    // Interface, handles SDL Renderer
+    class IRenderer {
+    public:
+        virtual ~IRenderer() = default;
+        // Size
+        virtual std::tuple<size_t, size_t> get_output_size() const = 0;
+        // Modifiers
+        virtual IRenderer & clear() = 0;
+        virtual IRenderer & present() = 0;
+        virtual IRenderer & copy(std::unique_ptr<ITexture> & texture, const Rect *source, const Rect *dest) = 0;
```
Comment (Drone29) [inline src/sdl_interface.h:68]: unique_ptr or shared_ptr ?

File: src/sdl_interface.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        // Texture
+        virtual std::unique_ptr<ITexture> create_texture(ISurface &surface) = 0;
+        virtual std::unique_ptr<ITexture> create_texture(uint32_t format, int access, int width, int heigth) = 0;
+    };
+
+    // Interface, handles SDL Texture
+    class ITexture {
+    public:
+        virtual ~ITexture() = default;
+        virtual std::tuple<uint32_t, int, int, int> query() const = 0;
```
Comment (Drone29) [inline src/sdl_interface.h:79]: is tuple ok here? or should query take a pointer to struct?

File: src/sdl_renderer.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    IRenderer& Renderer::present()
+    {
+        SDL_RenderPresent(m_renderer);
+        return *this;
+    }
+
+    IRenderer& Renderer::copy(std::unique_ptr<ITexture>& texture, const Rect* source, const Rect* dest)
+    {
+        auto tex = dynamic_cast<Texture*>(texture.get());
```
Comment (Drone29) [inline src/sdl_renderer.cpp:52]: couldn't come up with anything better than dynamic_cast here. ant ideas are welcome

File: src/sdl_window.cpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    IWindow &Window::restore()
+    {
+        SDL_RestoreWindow(m_window);
+        return *this;
+    }
+
+    std::unique_ptr<IRenderer> Window::create_renderer(int index, uint32_t flags)
+    {
+        auto rend = new Renderer(m_window, index, flags);
```
Comment (Drone29) [inline src/sdl_window.cpp:183]: Renderer has a private ctor, so no make_unique here

File: src/sdl_interface.h (hunk 3 above)
Comment (MikhailBerezhanov) [reply src/sdl_interface.h]: struct with named fields instead of tuple is better

**Comment by Drone29** [comment]
closing due to revision

**Pull request closed without merging**